docs/writing_a_simple_publisher_and_subscriber.md

```diff
@@ -0,0 +1,474 @@
+        })
+    }
+    fn data_callback(&self) -> Result<(), RclrsError> {
+        if let Some(data) = self.data.lock().unwrap().as_ref() {
+            println!("{}", data.data);
+        } else {
+            println!("No message available yet.");
+        }
+        Ok(())
+    }
+}
+fn main() -> Result<(), RclrsError> {
+    let context = Context::new(env::args()).unwrap();
+    let subscription = Arc::new(SimpleSubscriptionNode::new(&context).unwrap());
+    let subscription_other_thread = Arc::clone(&subscription);
+    thread::spawn(move || -> () {
+        iter::repeat(()).for_each(|()| {
+            thread::sleep(Duration::from_millis(1000));
+            subscription_other_thread.data_callback().unwrap()
+        });
+    });
+    rclrs::spin(subscription.node.clone())
+}
+```
+<details><summary>Examining the code in detail:</summary>
+
+#### SimpleSubscriptionNode:
+```rust
+pub struct SimpleSubscriptionNode {
+    node: Arc<Node>,
+    _subscriber: Arc<Subscription<StringMsg>>,
+    data: Arc<Mutex<Option<StringMsg>>>,
+}
+```
+Instead of a Publisher, there is a Subscription object in the Subscriber node. The data needs to be an `Arc<Mutex<Option<StringMsg>>>` because there can be errors in the data transfer process and this can be caught by including the value of the incoming subscription in an optional.  
+#### impl SimpleSubscriptionNode
+This code defines methods for the `SimpleSubscriptionNode` `struct`.  
+
+##### new
+The `new` method creates a ROS 2 node, subscriber, and a storage location for received messages, storing them in the `struct`.
+```rust
+    fn new(context: &Context) -> Result<Self, RclrsError> {
+        let node = create_node(context, "simple_subscription").unwrap();
+        let data: Arc<Mutex<Option<StringMsg>>> = Arc::new(Mutex::new(None));
+        let data_mut: Arc<Mutex<Option<StringMsg>>> = Arc::clone(&data);
+        let _subscriber = node
+            .create_subscription::<StringMsg, _>(
+                "publish_hello",
+                QOS_PROFILE_DEFAULT,
+                move |msg: StringMsg| {
+                    *data_mut.lock().unwrap() = Some(msg);
+                },
+            )
+            .unwrap();
+        Ok(Self {
+            node,
+            _subscriber,
+            data,
+        })
+    }
+
+```
+A few special features:  
+1. Initializing Shared Data:  
+    * `let data: Arc<Mutex<Option<StringMsg>>> = Arc::new(Mutex::new(None));`  
+        This line creates a shared data structure that will hold the received message.  
+        * `Arc<Mutex<Option<StringMsg>>>`: This is a complex type combining several functionalities:  
+            * `Arc<T>`: An atomically reference-counted pointer ([`Arc`](https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://doc.rust-lang.org/std/sync/struct.Arc.html&ved=2ahUKEwiJz_n3876FAxXZg_0HHc-yDZ8QFnoECAYQAQ&usg=AOvVaw2ZAPxD2olFejU3a_Ngb4f5)) allows multiple parts of the code to safely access the same data (`T`).
```

Comment:
Please use the real links from https://doc.rust-lang.org instead of Google links.

Reply:
done